README


This archive contains the distribution AnyEvent-Mojo, version
0.8003:

  Start async Mojo servers easly

This software is copyright (c) 2009 by Pedro Melo.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Changes
Revision history for AnyEvent-Mojo

0.8003 Sat Oct 10 17:05:18 UTC 2009
         - Disable the HTTP proxy some CPAN testers use: connections are for 127.0.0.1!
         - Improve test log messages

0.8002 Sat Oct 10 12:15:29 UTC 2009
         - Improve test messages to figure out CPAN Testers failures
         - Remove extra debug from ::Connection

0.8001 Sun Oct  4 14:12:23 UTC 2009
         - Fixed RT #46595 (reported by JWRIGHT)
         
0.8    Sun Oct  4 14:08:49 UTC 2009
         - Rewrote Connection to support latest Mojo with Pipeline support
         - Move all AnyEvent stuff to server, Connection is AnyEvent-free
         - resume() method is disabled for now: need to workaround Mojo
         - started using Dist::Zilla to manage this distribution

0.6002 Tue Nov 18 16:52:52 WET 2008
         - D'oh: fix pause/resume and timeouts

0.6001 Thu Nov  6 16:11:23 WET 2008
         - Don't close the socket, it might be too soon

0.6    Thu Nov  6 15:27:32 WET 2008
         - New functional interface to AnyEvent::Mojo, mojo_server()
         - Updated server docs
         - Renamed AnyEvent::Mojo to AnyEvent::Mojo::Server
         - refactor test code

0.5    Tue Nov  4 11:19:10 WET 2008
         - Added pause()/resume() methods to our Connection class
         - Remove the dependency on Mojo 0.7
         - Fix: Latest Mojo::Base no longer accepts multiple argumets on attributes
         - Fix Content-Length handling

0.4    Mon Oct 27 08:26:37 WET 2008
         - Brown-paper-bag mistake: wrong Mojo version required

0.3    Sun Oct 26 20:38:44 WET 2008
         - Move the test server, MyTestServer, to a tlib directory
         - Cope with Mojo 0.07

0.2    Sat Oct 25 13:31:38 WEST 2008
         - Added support for keep-alive requests to Connection
         - Make sure we don't die on SIGPIPE's
         - The startup banner is printed after the sucessful startup of the server now
         - Default timeout reduced to 5 seconds
         - Added stop() method to stop a server
         - Improve documentation and tests

0.1    Tue Oct 21 18:00:03 WEST 2008
         - First working version
         - Supports HTTP/1.1 without keep-alive




t/tlib/MyTestServer.pm
package MyTestServer;
our $VERSION = '0.8003';



use strict;
use warnings;
use AnyEvent;
use base 'AnyEvent::Mojo::Server';
use AnyEvent::HTTP;
use Test::More;

{
  ## Thou shall not use a proxy for localhost
  no warnings;
  undef $AnyEvent::HTTP::PROXY;
}

BEGIN {
  __PACKAGE__->attr('banner_called');
}

sub startup_banner {
  my $self = shift;
  
  $self->banner_called([$self->host, $self->port]);
  return $self->SUPER::startup_banner(@_);
}


###################################
# Fork off a server for us, please?

sub start_server {
  my $class = shift;
  my $port = shift;
  my $cb = pop;
  my %args = @_;

  # Default port: random
  $port ||= 4000 + $$ % 10000;
    
  my $pid = fork();
  die "Could not fork: $!" unless defined $pid;

  # Parent
  if ($pid) {
    diag("Waiting for server to start...");
     
    my $count = 0;
    my $done = AnyEvent->condvar;
    my $checker; $checker = AnyEvent->timer(
      after => 1,
      interval => 2,
      cb => sub {
        AnyEvent::HTTP::http_get(
          "http://127.0.0.1:$port/",
          timeout => 1,
          cb => sub {
            if (!defined($_[0])) {
              $count++;
              die "Could not start test server!\n" if $count > 10;
              return;
            }
            undef $checker;
            $done->send;
          },
        );
      },
    );
    
    $done->recv; # Wait for the server to start...
    
    return wantarray? ($pid, $port) : $pid;
  }


  # Child
  my $server = AnyEvent::Mojo::Server->new;
  $server->keep_alive_timeout($args{keep_alive_timeout})
      if defined $args{keep_alive_timeout};
  $server->port($port)->handler_cb($cb);
  
  $server->run;
  exit(0);
}

sub stop_server {
  my ($class, $pid) = @_;
  
  return unless $pid;
  
  is(kill(15, $pid), 1, "killed one"); # sigterm
  is(waitpid($pid, 0), $pid, "waipid ok");

  return;
}

1;

lib/AnyEvent/Mojo.pm
package AnyEvent::Mojo;
our $VERSION = '0.8003';



use strict;
use warnings;
use 5.008;
use AnyEvent::Mojo::Server;
use Carp qw( croak );
use base qw( Exporter );

@AnyEvent::Mojo::EXPORT = qw( mojo_server );


#####################
# Start a Mojo server

sub mojo_server {
  my %args;
  if (ref($_[0]) eq 'HASH') {
    %args = %{$_[0]};
  }
  else {
    @args{qw( host port handler_cb )} = @_;
  }
  
  croak('FATAL: the handler callback is required, ')
    unless ref($args{handler_cb}) eq 'CODE';
  
  my $server = AnyEvent::Mojo::Server->new(%args);
  
  $server->listen;
  
  return $server;
}


42; # End of AnyEvent::Mojo

__END__

=encoding utf8

=head1 NAME

AnyEvent::Mojo - Start async Mojo servers easly


=head1 VERSION

version 0.8003

=head1 SYNOPSIS

    use strict;
    use warnings;
    use AnyEvent;
    use AnyEvent::Mojo;
    
    my $port = 7865;
    
    my $server = mojo_server undef, $port, sub {
      my ($self, $tx) = @_;
      
      # Handle the request here, see AnyEvent::Mojo::Server for details
    };
    
    # or...

    my $server = mojo_server {
      host       => '127.0.0.1',
      port       => $port,
      handler_cb => sub {
          my ($self, $tx) = @_;
          
          # Handle the request here, see AnyEvent::Mojo::Server for details
      },
    };    
        
    # Run the loop
    $server->run
    
    # ... or ...
    AnyEvent->condvar->recv;


=head1 STATUS

This is a first B<beta> release. The interface B<should not> change
in a backwards incompatible way until version 1.0.



=head1 DESCRIPTION

This module allows you to integrate Mojo applications with the AnyEvent
framework. For example, you can run a web interface for a long-lived
AnyEvent daemon.


=head1 FUNCTIONS

The module exports the following functions:

=head2 mojo_server

Starts a server. Accepts three parameters:

=over 4

=item host

The hostname or IP address to which the server will bind to. Use C<undef> to
bind to all interfaces.


=item port

Port where the server will listen on. You can use C<undef> to choose the
default value of 3000.


=item cb

A coderef. This handler will be called for each request. The first parameter
is the server object, and the second is a C<Mojo::Transaction>.

=back

Alternatively you can pass a hash or hashref with all the options that will
be passed along to C<AnyEvent::Mojo::Server> constructor.

Returns a C<AnyEvent::Mojo::Server> object.



=head1 SEE ALSO

L<AnyEvent::Mojo::Server>, L<Mojo>, and L<AnyEvent>.



=head1 AUTHOR

Pedro Melo, C<< <melo at cpan.org> >>



=head1 CREDITS

In alphabetical order:

Lars Balker Rasmussen

Sebastian Riedel



=head1 COPYRIGHT & LICENSE

Copyright 2008-2009 Pedro Melo.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

lib/AnyEvent/Mojo/Server/Connection.pm
package AnyEvent::Mojo::Server::Connection;
our $VERSION = '0.8003';



use strict;
use warnings;
use parent 'Mojo::Base';
use Mojo::Transaction::Pipeline;
use Carp;

__PACKAGE__->attr([qw( server pipeline write_mode_cb close_sock_cb )]);
__PACKAGE__->attr([qw( remote_address remote_port local_address local_port )]);
__PACKAGE__->attr(request_count => 0);


##############################
# Close the current connection

sub close {
  my ($self, $type, $mesg) = @_;

  $self->pipeline(undef);
  $self->write_mode_cb(undef);
  $self->server(undef);

  ## Make sure the socket is closed
  my $cb = $self->close_sock_cb;
  $self->close_sock_cb(undef);
  $cb->();

  return;
}


##########################################
# Pause/restart processing of transactions

sub pause {
  my ($self) = @_;
  my $p = $self->pipeline;
  my $tx = $p && $p->server_tx;
  
  croak("pause() only works on tx's in the 'handle_request' state")
    unless $tx && $tx->is_state('handle_request');
  
  my ($tx_state, $tx_req_state) = ($tx->state, $tx->req->state);
  $tx->state('paused');
  $tx->req->state('paused');
  
  return sub { $self->_resume($tx_state, $tx_req_state) };
}

sub resume {
  Carp::croak('resume() not available, use the cb that pause() gave you, ')
}

sub _resume {
  my ($self, $tx_state, $tx_req_state) = @_;
  my $p = $self->pipeline;
  my $tx = $p && $p->server_tx;

  return unless $tx && $tx->is_state('paused');

  $tx->req->state($tx_state);
  $tx->state($tx_state);

  $p->server_handled;
  $p->server_spin;
  $self->_check_for_writters;
  
  return;
}


###############
# _on_ handlers

sub _on_read {
  my ($self, $buf) = @_;
  my $p = $self->pipeline;
  my $srv = $self->server;

  # Make sure we have a pipeline
  $p = $self->_mk_pipeline unless $p;

  my $tx = $p->server_tx;
  while ($buf) {
    # Need a new transaction?
    unless ($tx) {
      $tx = $srv->build_tx_cb->($srv);
      $p->server_accept($tx);
    }

    $p->server_read($buf);
    
    if ($p->is_state('handle_continue')) {
      $srv->continue_handler_cb->($srv, $p->server_tx);
      $p->server_handled;
    }

    if ($p->is_state('handle_request')) {
      my $tx = $p->server_tx;
      
      $srv->handler_cb->($srv, $tx);
      $p->server_handled unless $tx->state eq 'paused';
    }
    
    $p->server_spin;
    
    $buf = $p->server_leftovers;
    $tx = undef if $buf;
  }
  
  $self->_check_for_writters;
  $self->_cleanup;
}

sub _on_write {
  my ($self, $write_cb) = @_;

  return unless my $p = $self->pipeline;
  return unless my $chunk = $p->server_get_chunk;

  ## FIXME: The order of is wrong. We should call server_written with
  ## the output of $write_cb->(chunk)
  # but the problem is that the AnyEvent impl of AnyEvent::Handle
  # recursivelly calls us again when we call $write_cb. So we need to
  # move the state machine before we call it. Hence the call to
  # $p->server_spin.
  #
  # I don't have a better solution right now. One possibility would be
  # not to use AnyEvent::Handle, but that would mean to rewrite a lot of
  # tricky code. Another would be to use the autocork feature of
  # AnyEvent that delays the write until the next I/O loop iteraction.
  my $written = length($chunk);
  $p->server_written($written);
  $self->_cleanup;
  
  $write_cb->($chunk);
}

sub _on_error {
  my ($self, undef, $mesg) = @_;
  
  $self->close('on_error', $mesg)
}

sub _on_eof {
  my ($self) = @_;
  
  $self->close('on_error')
}

sub _on_timeout {
  my $self = shift;
  my $p = $self->pipeline;
  my $tx = $p && $p->server_tx;

  return $self->close('timeout') unless $tx && $tx->state eq 'paused';
  return;
}


##############
# Spin the web

sub _check_for_writters {
  my $self = shift;
  my $p = $self->pipeline;
  
  ## Start the writer if we are ready for it
  $self->write_mode_cb->($p->server_is_writing);
}

sub _cleanup {
  my $self = $_[0];
  my $p = $self->pipeline;
  return unless $p;
  
  $p->server_spin;

  if ($p->is_finished) {
    $self->pipeline(undef);
    $self->close('no-keep-alive') unless $p->keep_alive;
  }
}

sub _mk_pipeline {
  my ($self) = @_;
  
  my $srv = $self->server;
  my $p = Mojo::Transaction::Pipeline->new;
  $p->connection($self);
  $p->kept_alive(1) if $self->_inc_request_count > 1;

  # Store connection information
  $p->local_address($self->local_address);
  $p->local_port($self->local_port);
  $p->remote_address($self->remote_address);
  $p->remote_port($self->remote_port);
  
  $self->pipeline($p);
  
  return $p;
}


#######
# Stats

sub _inc_request_count {
  my $self = $_[0];
  $self->server->_inc_request_count;
  return ++$self->{request_count}
}

42; # End of AnyEvent::Mojo::Server::Connection

__END__

=encoding utf8

=head1 NAME

AnyEvent::Mojo::Server::Connection - An active TCP connection to AnyEvent::Mojo::Server



=head1 VERSION

version 0.8003

=head1 SYNOPSIS

    use AnyEvent::Mojo::Server::Connection;

    ...


=head1 DESCRIPTION

Foreach connection to a L< AnyEvent::Mojo::Server >,
a C< AnyEvent::Mojo::Server::Connection > object is created.

This object keeps track of the current L< Mojo::Transaction >.

If an error or EOF condition is detected while reading or writting to the
client socket, or in case of a timeout, the socket is disconnected.


=head1 METHODS

=head2 new

The constructor accepts the following parameters:


=over 4

=item sock

The client socket.


=item peer_host

The IP of the client.


=item peer_port

The TCP port number of the client.


=item server

The L< AnyEvent::Mojo::Server > to whom this connection belongs to.


=item timeout

Seconds (can be fractional) that the connection can be idle (waiting
for a request or unable to write more data out).

If the connection is paused, the timeout is ignored.

If the timeout fires, the connection is closed.


=back


It returns the C< AnyEvent::Mojo::Server::Connection > object.



=head2 run

The C< run() > method starts all the L< AnyEvent::Handle > processing to read
the next request, process it and write the response.

It returns nothing.



=head2 close

The C< close() > method clears the current transaction, destroys the 
L< AnyEvent::Handle > associated with this connection and closes the
client socket.


=head2 request_count

Returns the total request count for the connection. In case of keep-alive
requests, the request count grows beyond 1. 


=head2 peer_host

Returns the IP address of the peer host.


=head2 peer_port

Returns the TCP port number of the peer host.


=head1 ASYNCHRONOUS PROCESSING

While in the middle of a request, an application can pause the current
transaction, do something else (including dealing with other requests)
and then resume the processing.

To do that, you application must call the C<$tx->connection->pause()> method.

When you are ready to send back the response, call
C<$tx->connection->resume()>.

For example:

    # inside your response handler of you Mojo::App
    $tx->connection->pause();
    
    # Call webservice and deal with result
    http_get 'http://my.webservice.endpoint/api', sub {
      my ($data) = @_;
      
      $tx->response->body("Webservice returned this: '$data'");
      $tx->connection->resume();
    };

To make it easier to resume later, the C<pause()> method returns a coderef
that will resume the transaction when called. So the code above could be
written like this:

    # inside your response handler of you Mojo::App
    my $resume_cb = $tx->connection->pause();
    
    # Call webservice and deal with result
    http_get 'http://my.webservice.endpoint/api', sub {
      my ($data) = @_;
      
      $tx->response->body("Webservice returned this: '$data'");
      $resume_cb->();
    };



=head2 pause()

Pauses the current transaction.

The transaction state must be C<write>, that is, before sending any status
or header responses.

While the connection is paused, inactivity timeouts are ignored.

Returns a coderef that, when called, will resume the transaction.


=head2 resume()

Resumes a paused transaction.

The response must be complete and we will immediatly start sending the data
to the client.

Returns nothing.


=head1 AUTHOR

Pedro Melo, C<< <melo at cpan.org> >>



=head1 COPYRIGHT & LICENSE

Copyright 2008-2009 Pedro Melo.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

lib/AnyEvent/Mojo/Server.pm
package AnyEvent::Mojo::Server;
our $VERSION = '0.8003';



use strict;
use warnings;
use 5.008;
use parent 'Mojo::Server';
use Carp qw( croak );
use AnyEvent;
use AnyEvent::Socket;
use AnyEvent::Handle;
use AnyEvent::Mojo::Server::Connection;
use IO::Socket qw( SOMAXCONN );

__PACKAGE__->attr('host');
__PACKAGE__->attr(port => 3000);

__PACKAGE__->attr(listen_queue_size => sub { SOMAXCONN });
__PACKAGE__->attr(max_keep_alive_requests => 100);
__PACKAGE__->attr(keep_alive_timeout => 5);
__PACKAGE__->attr(request_count => 0);

__PACKAGE__->attr([qw( run_guard listen_guard)]);
__PACKAGE__->attr(connection_class => 'AnyEvent::Mojo::Server::Connection');


sub listen {
  my $self = shift;
  
  # Already listening
  return if $self->listen_guard;
  
  my $guard = tcp_server($self->host, $self->port,
    sub { $self->_on_connect(@_) },
    # Setup listen queue size, record our hostname and port
    sub {
      $self->host($_[1]);
      $self->port($_[2]);
      
      return $self->listen_queue_size;
    }
  );
  
  $self->listen_guard(sub { $guard = undef });
  $self->startup_banner;
  
  return;
}

sub _on_connect {
  my ($self, $sock, $remote_address, $remote_port) = @_;
  
  if (!$sock) {
    $self->log("Connect failed: $!");
    return;
  }
  
  my $con = $self->connection_class->new(
    local_address  => $self->host,
    local_port     => $self->port,
    remote_address => $remote_address,
    remote_port    => $remote_port,
    server         => $self,
  );
  
  my $hdl; $hdl = AnyEvent::Handle->new(
    fh         => $sock,
    timeout    => $self->keep_alive_timeout,

    on_read    => sub { $con->_on_read(delete $_[0]->{rbuf}) },
    on_eof     => sub { $con->_on_eof(@_);   $hdl->destroy if $hdl },
    on_error   => sub { $con->_on_error(@_); $hdl->destroy if $hdl },
    on_timeout => sub { $con->_on_timeout(@_) },
  );
  
  $con->write_mode_cb(sub {
    my $on = shift;
    
    if ($on) {
      # print STDERR "## [handle] SET ON DRAIN\n";
      $hdl->on_drain(sub {
        my $h = $_[0];
        $con->_on_write(sub {
          # print STDERR "## [handle] PUSH WRITE\n";
          $h->push_write($_[0]);
          return length($_[0]);
        });
      });
    }
    else {
      # print STDERR "## [handle] Remove on_drain\n";
      $hdl->on_drain(undef);
    }
  });
  $con->close_sock_cb(sub { $hdl = undef });
  
  return;
}

sub run {
  my $self = shift;
  
  $SIG{PIPE} = 'IGNORE';
  
  # Start the server socket
  $self->listen;
  
  # Create a run guard
  my $cv = AnyEvent->condvar;
  $self->run_guard(sub { $cv->send });

  $cv->recv;
  
  return;
}

sub stop {
  my ($self) = @_;
  
  # Clears the listening guard, closes the listening socket
  if (my $cb = $self->listen_guard) {
    $cb->();
    $self->listen_guard(undef);
  }
  
  # Clear the run() guard
  if (my $cb = $self->run_guard) {
    $cb->();
    $self->run_guard(undef);
  }
}

sub startup_banner {}

#######
# Stats

sub _inc_request_count {
  return ++$_[0]->{request_count}
}


42; # End of AnyEvent::Mojo::Server

__END__

=encoding utf8

=head1 NAME

AnyEvent::Mojo::Server - Run Mojo apps using AnyEvent framework



=head1 VERSION

version 0.8003

=head1 SYNOPSIS

    use strict;
    use warnings;
    use AnyEvent;
    use AnyEvent::Mojo::Server;
    
    my $server = AnyEvent::Mojo::Server->new;
    $server->port(3456)->listen_queue_size(10);
    $server->max_keep_alive_requests(100)->keep_alive_timeout(3);
    
    $server->handler_cb(sub {
      my ($self, $tx) = @_;
      
      # Do whatever you want here
      $you_mojo_app->handler($tx);

      # Cool stats
      $tx->res->headers(
        'X-AnyEvent-Mojo-Request-Count' =>  $server->request_count
      );
      
      return $tx;
    });
    
    # Start it up and keep it running
    $server->run
    
    # integrate with other AnyEvent stuff
    $server->listen
    
    # other AnyEvent stuff here
    
    # Run the loop
    AnyEvent->condvar->recv;
    
    # Advanced usage: use your own Connection class
    $server->connection_class('MyConnectionClass');


=head1 STATUS

This is a first B<beta> release. The interface B<should> not change in a
backwards incompatible way until version 1.0.

Basic HTTP/1.0 and HTTP/1.1 works, including keep-alives.

Support for 100-Continue is still missing.


=head1 DESCRIPTION

This module allows you to integrate Mojo applications with the AnyEvent
framework. For example, you can run a web interface for a long-lived
AnyEvent daemon.

The AnyEvent::Mojo::Server extends the Mojo::Server class.

To use you need to create a AnyEvent::Mojo::Server object. You can set the port
with the C< port() > method.

Then set the request callback with the Mojo::Server method, 
C<handler_cb()>.

This callback will be called on every request. The first parameter is
the L<AnyEvent::Mojo::Server> server object itself, and the second parameter is a
L<Mojo::Transaction>.

The code should build the response and return.

Optionally, you can pause the current request and resume it later. This
is useful if you need to request information from a remote service to
answer the current request. You pause the current Mojo request, ask the
information asyncronously, and resume when the answer arrives. In the
meantime, you can keep answering other requests.



=head1 METHODS


=head2 new

Creates a new L<AnyEvent::Server::Mojo> instance.

Accepts an hash/hashref with options. All the following methods can be used
as options. Check the method documentation below for valid values and an
explanation about each one.

=over 4

=item host

=item port

=item listen_queue_size

=item max_keep_alive_requests

=item keep_alive_timeout

=item connection_class

=back


Returns a L<AnyEvent::Server::Mojo> object.


=head2 host

The hostname or IP address to which the server will bind to. By default it
will bind to all interfaces.


=head2 port

Port where the server will listen on. Defaults to 3000.


=head2 listen_queue_size

Defines the size of the listening queue. Defaults to C< SOMAXCONN >.

Use

    perl -MSocket -e 'print Socket::SOMAXCONN,"\n"'

to discover the default for your operating system.


=head2 max_keep_alive_requests

Number of requests that each connection will allow in keep-alive mode.

Use 0 for unlimited requests. Default is 100 requests.


=head2 keep_alive_timeout

Number of seconds (can be fractional) that the server lets open connections
stay idle.

Default is 5 seconds.


=head2 request_count

Returns the number of requests the server has answered since it started.


=head2 connection_class

Sets the class name that will be used to process each connection.

Defaults to L< AnyEvent::Mojo::Server::Connection >.


=head2 listen

Starts the listening socket.

Returns nothing.


=head2 run

Starts the listening socket and kickstarts the
L< AnyEvent > runloop.


=head2 stop

Closes the listening socket and stops the runloop initiated by a call to
C< run() >.


=head2 startup_banner

Called after the listening socket is started. You can override this method
on your L< AnyEvent::Mojo::Server > subclasses to setup other components.

The default C< startup_banner > is empty.



=head1 AUTHOR

Pedro Melo, C<< <melo at cpan.org> >>



=head1 COPYRIGHT & LICENSE

Copyright 2008-2009 Pedro Melo.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.